common/thrift_client_pool.h

```diff
@@ -307,24 +307,30 @@ class ThriftClientPool {
     }
 
     void cleanupStaleChannels(const folly::SocketAddress& addr) {
-      // cleanup stale entries if it hasn't been done for a period of time.
       auto now = time(nullptr);
-      if (last_cleanup_time_ + FLAGS_channel_cleanup_min_interval_seconds
-          < now) {
-        last_cleanup_time_ = now;
-        auto itor = channels_.find(addr);
-        int n = 0;
-        while (itor != channels_.end() &&
-               n++ < FLAGS_channel_max_checking_size) {
-          // We don't cleanup !good() live channels here. Otherwise we
-          // will need to upgrade it to a shared_ptr. We expect clients
-          // won't keep a !good() channels for a long period of time.
-          if (itor->second.first.use_count() == 0) {
-            itor = channels_.erase(itor);
-          } else {
-            ++itor;
-          }
-        }
+      // skip cleanup if it was done recently
+      if (now < last_cleanup_time_ + FLAGS_channel_cleanup_min_interval_seconds) {
+	return;
+      }
+
+      last_cleanup_time_ = now;
+      auto itor = channels_.find(addr);
+      int n = 0;
+      while (itor != channels_.end() &&
+	     n++ < FLAGS_channel_max_checking_size) {
+	auto c = itor->second.first.lock();
+	// the channel has been released
```

Comment:
Q: under what circumstances will the channel be released? (or put it another way, clarify what "release" means here, which is less clear than "closed")

Reply:
Released means all shared_ptrs that are associated with the weak_ptr has been destructed. So the underlying channel object has been destructed and its connection has been closed.

There are two layers: thrift_client_pool and thrift_router. 

In the initial design of thrift_client_pool, the assumption was that its client (thrift_router, etc) will release (destruct) the returned client shared_ptr object shortly after the connection becomes bad (such as closed by the remote peer). This assumption turned out to be not always true. thrift_router stores the shared_ptr client objects from thrift_client_pool in a thread local data structure. Though active threads will release (destruct) the shard_ptr objects shortly after a connection becomes bad, idle thread won't. (Some services have hundreds of worker threads, and they are scheduled in a LIFO way, i.e., the worker pool always tries to use the idle threads that were most recently used).

Reply:
thanks for the detailed explanation!